src/bgg/sampler.rs

```diff
@@ -100,36 +99,43 @@ where
         error_sampler: S,
         gauss_sigma: f64,
     ) -> Self {
-        // s_init := (sampled secret, -1)
-        let minus_one_poly = <S::M as PolyMatrix>::P::const_minus_one(params);
-        let mut secrets = secrets.to_vec();
-        secrets.push(minus_one_poly);
-        let secret_vec = S::M::from_poly_vec_row(params, secrets);
-        Self { secret_vec, error_sampler, gauss_sigma }
+        Self {
+            secret_vec: S::M::from_poly_vec_row(params, secrets.to_vec()),
+            error_sampler,
+            gauss_sigma,
+        }
     }
 
     /// This extend the given plaintexts +1 and insert constant 1 polynomial plaintext
     /// Actually in new simplified construction, this sample is not used unless debug
+    ///
+    /// use_structured_error for sample BGGEncoding for ABE context: Algorithm 4.https://eprint.iacr.org/2017/601.pdf
     pub fn sample(
         &self,
         params: &<<<S as PolyUniformSampler>::M as PolyMatrix>::P as Poly>::Params,
         public_keys: &[BggPublicKey<S::M>],
         plaintexts: &[<S::M as PolyMatrix>::P],
+        use_structured_error: bool,
```

Comment:
Instead of `use_structured_error`, can you add a flag `add_error`, which adds independent errors to the sampled encoding if it is true (for diamond-io); otherwise adds no error (for abe)?